### ISSUE TITLE
Convert symbolic fermions to Majoranas

### ISSUE THREAD

**Issue by williamesamuelson** [cvsvensson/QuantumDots.jl#140]
Would it make sense to implement a function to convert an expression of symbolic fermions into symbolic Majoranas and vice versa?

I tried
```julia
@fermions c
@majoranas a b
function ferm2maj_sub(a, b, op)
    i = op.label
    return op.creation ? (op => 1/2*(a[i] + 1im*b[i])) : (op => 1/2*(a[i] - 1im*b[i])) 
end
ops = mapreduce(i -> [c[i], c[i]'], vcat, 1:2)
fermion_maj_sub = Dict(ferm2maj_sub(a, b, op) for op in ops)
```
which then could be used in `substitute`.

**Comment by cvsvensson** [comment]
Yes, that would make sense. We have to decide on a convention for how to pair up majoranas into fermions. I guess the way you do it makes sense. Two different species of Majoranas make up a fermion. 

The drawback is that if I work with a single species of Majoranas, and then decide I want to convert it to fermions, I'd be in trouble. Either I'd have to go back and redo all calculations with two different species, or use twice as many fermions. Or we could provide a way to combine Majoranas with different labels into a fermion. 

We could enforce using two different species of Majoranas. Writing `@majoranas γ` could define two species, γ and γ̃. The advantage is that there is a single correct choice that we enforce. The disadvantage is that I like the simplicity of working with a single species sometimes, and we have to pick a nice name for the second species.

I'm leaning towards having two functions. 
`f(γ, γ̃, c)` returns something that converts two majorana species into a fermion species `c`
`f(γ, labelpairs, c)` returns something that converts one majorana species into a fermion species `c`, pairing up labels according to `labelpairs`. I'm not sure what `labelpairs` should be. Could be a list or dict of pairs, but then the user has to actually write down all the labels they use. Perhaps it should be a function which is its own inverse. This function would map from one label to the other.

As for how to actually implement it, I would take a look at using `@rule`, see https://docs.sciml.ai/SymbolicUtils/stable/manual/rewrite/. I tested it here:
https://github.com/cvsvensson/QuantumDots.jl/blob/60e63bff83302deec2d466331347f1fb948b02f2/src/symbolics/symbolic_majoranas.jl#L139-L140
This rule goes in and changes the labels of the fermions. It maps `c[k]` to `c[k+1]`. Then we avoid having to create a large dictionary for the substitution. We just make a rule that converts between majoranas and fermions. Just like `substitute` this should in principle work automatically for additions and multiplications of several fermions. This rule would be what is returned by the function `f` above.

**Comment by cvsvensson** [comment]
I was wrong about `@rule` working on subexpressions. But there is a simple solution, https://github.com/JuliaSymbolics/Symbolics.jl/issues/356.
See tests:
https://github.com/cvsvensson/QuantumDots.jl/blob/cb830d732223a643cf911eeaa39d76e72db4db71/src/symbolics/symbolic_majoranas.jl#L139-L146

**Comment by williamesamuelson** [comment]
Looks nice, thanks! I'll give it a shot.

**Comment by williamesamuelson** [comment]
#142 partially implements this. However, it would be nice to have support for using one Majorana species, so I'll keep this open.
